crates/turborepo-lib/src/run/summary/execution.rs

```diff
@@ -375,114 +368,96 @@ impl TaskTracker<chrono::DateTime<Local>> {
+    pub async fn cached(self) -> TaskExecutionSummary {
         let Self {
             sender,
             started_at,
             task_id,
         } = self;
 
         let ended_at = Local::now();
-        let execution = Some(TaskExecutionSummary {
+        let execution = TaskExecutionSummary {
             start_time: started_at.timestamp_millis(),
             end_time: ended_at.timestamp_millis(),
-            state: ExecutionState::Cached,
-            exit_code: None,
-        });
-
-        sender
-            .send(TrackerMessage::Finished(TaskState { task_id, execution }))
-            .await
-            .expect("summary state thread finished");
-    }
+            // Go synthesizes a zero exit code on cache hits
+            exit_code: Some(0),
+            error: None,
+        };
 
-    pub async fn build_succeeded(self, exit_code: i32) {
-        let Self {
-            sender,
-            started_at,
+        let state = TaskState {
             task_id,
-        } = self;
-
-        let ended_at = Local::now();
-        let execution = Some(TaskExecutionSummary {
-            start_time: started_at.timestamp_millis(),
-            end_time: ended_at.timestamp_millis(),
-            state: ExecutionState::Built { exit_code },
-            exit_code: Some(exit_code),
-        });
-
+            execution: Some(execution.clone()),
+        };
         sender
-            .send(TrackerMessage::Finished(TaskState { task_id, execution }))
+            .send(TrackerMessage {
+                event: Event::Cached,
+                state: Some(state),
+            })
             .await
             .expect("summary state thread finished");
+        execution
     }
 
-    pub async fn build_failed(self, exit_code: i32, error: impl fmt::Display) {
+    pub async fn build_succeeded(self, exit_code: i32) -> TaskExecutionSummary {
         let Self {
             sender,
             started_at,
             task_id,
         } = self;
 
         let ended_at = Local::now();
-        let execution = Some(TaskExecutionSummary {
+        let execution = TaskExecutionSummary {
             start_time: started_at.timestamp_millis(),
             end_time: ended_at.timestamp_millis(),
-            state: ExecutionState::BuildFailed {
-                exit_code,
-                err: error.to_string(),
-            },
             exit_code: Some(exit_code),
-        });
+            error: None,
+        };
 
+        let state = TaskState {
+            task_id,
+            execution: Some(execution.clone()),
+        };
         sender
-            .send(TrackerMessage::Finished(TaskState { task_id, execution }))
+            .send(TrackerMessage {
+                event: Event::Built,
+                state: Some(state),
+            })
             .await
             .expect("summary state thread finished");
+        execution
     }
 
-    pub async fn spawn_failed(self, error: impl fmt::Display) {
+    pub async fn build_failed(
```

Comment:
Move ourselves closer to Go behavior where spawn failures and non-zero exit codes were treated the same.